### PR TITLE
minSdk 24

### PR DESCRIPTION
**Pull request by mikehardy**
## Purpose / Description

Bumps our minSdk to 24, the commits have a bit more info

## Approach

bump the SDK
inline the compat stuff I can inline
inline the resources that are now mainline vs v24
remove the version checks no longer needed

## How Has This Been Tested?

Just keep running gradle targets (`./gradlew lint`) is especially helpful as it checks unnecessary SDK checks

## Learning (optional, can help others)

coding is a good distraction

## Checklist

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/TtsVoices.kt [hunk 1]
```diff
@@ -280,8 +283,56 @@ class AndroidTtsVoice(val voice: Voice, val engine: String) : TtsVoice(name = "$
...
+            } catch (e: Exception) {
+                Timber.w(e, "Failed to normalize locale %s", locale)
+                locale
+            }
+        }
+
+        /**
+         * Maps from the ISO 3 code of a locale to the locale in
+         */
+        private val twoLetterSystemLocaleMapping: Map<String, Locale>
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/TtsVoices.kt:315]: Add a TODO to move these into a separate object - obvious coupling between the methods, the `init {}` block and this field

**Review by david-allison — CHANGES_REQUESTED** [review]
All looks good.

Patch contains various minor nitpicks as I was reviewing

One request with the `Shortcut[Group]`, as that feels big enough for its own commit, and the code feels too heavy for `AnkiActivity`

```patch
Subject: [PATCH] 
---
Index: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/AudioRecorder.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/AudioRecorder.kt b/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/AudioRecorder.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/AudioRecorder.kt	(revision a0c8689b193165f444da968a1031d37b9c92b144)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/AudioRecorder.kt	(date 1730961490260)
@@ -103,15 +103,12 @@
     }
 
     fun pause() {
-        if (!this::recorder.isInitialized) {
-            return
-        }
+        if (!this::recorder.isInitialized) return
         recorder.pause()
     }
 
     fun resume() {
-        if (this::recorder.isInitialized) {
-            recorder.resume()
-        }
+        if (!this::recorder.isInitialized) return
+        recorder.resume()
     }
 }
Index: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/ModelEditorContextMenu.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/ModelEditorContextMenu.kt b/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/ModelEditorContextMenu.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/ModelEditorContextMenu.kt	(revision a0c8689b193165f444da968a1031d37b9c92b144)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/ModelEditorContextMenu.kt	(date 1730961490255)
@@ -23,8 +23,7 @@
     @SuppressLint("CheckResult")
     override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
         super.onCreate(savedInstanceState)
-        var availableItems = ModelEditorContextMenuAction.entries.toList()
-        availableItems = availableItems.sortedBy { it.order }
+        val availableItems = ModelEditorContextMenuAction.entries.sortedBy { it.order }
 
         return AlertDialog.Builder(requireActivity()).create {
             setTitle(requireArguments().getString(KEY_LABEL))
Index: AnkiDroid/src/androidTest/java/com/ichi2/anki/TtsVoicesTest.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/androidTest/java/com/ichi2/anki/TtsVoicesTest.kt b/AnkiDroid/src/androidTest/java/com/ichi2/anki/TtsVoicesTest.kt
--- a/AnkiDroid/src/androidTest/java/com/ichi2/anki/TtsVoicesTest.kt	(revision a0c8689b193165f444da968a1031d37b9c92b144)
+++ b/AnkiDroid/src/androidTest/java/com/ichi2/anki/TtsVoicesTest.kt	(date 1730961490275)
@@ -17,7 +17,6 @@
 package com.ichi2.anki
 
 import androidx.test.ext.junit.runners.AndroidJUnit4
-import com.ichi2.anki.AndroidTtsVoice.Companion.normalize
 import com.ichi2.anki.tests.InstrumentedTest
 import org.hamcrest.CoreMatchers.equalTo
 import org.hamcrest.MatcherAssert.assertThat
@@ -30,7 +29,7 @@
     @Test
     fun normalize() {
         fun assertEqual(l: Locale, str: String) {
-            val normalized = normalize(l)
+            val normalized = AndroidTtsVoice.normalize(l)
             assertThat(normalized.toLanguageTag(), equalTo(str))
         }
 
Index: AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt b/AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt	(revision a0c8689b193165f444da968a1031d37b9c92b144)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt	(date 1730962022591)
@@ -472,6 +472,7 @@
     }
 
     private fun localeHintDialog() {
+        Timber.i("displaying locale hint dialog")
         // We don't currently show the current value, but we may want to in the future
         val dialogFragment: DialogFragment = LocaleSelectionDialog.newInstance(this)
         showDialogFragment(dialogFragment)

```

File: AnkiDroid/src/main/java/com/ichi2/compat/CompatV24.kt [hunk 2]
```diff
@@ -16,162 +16,248 @@
...
+import java.io.IOException
+import java.io.InputStream
+import java.io.OutputStream
+import java.io.Serializable
+import kotlin.time.Duration
+
+/** Baseline implementation of [Compat]. Check [Compat]'s for more detail.  */
+@KotlinCleanup("add extension method logging file.delete() failure" + "Fix Deprecation")
+@Suppress("Deprecation")
+open class CompatV24 : Compat {
```
[the remark this answers is not in the document]
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/compat/CompatV24.kt:50]: I like that idea, thank you!
That combined with not re-homing all the shortcut stuff into AnkiActivity will make it pretty reasonable

File: AnkiDroid/src/main/java/com/ichi2/anki/TtsVoices.kt (hunk 1 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/TtsVoices.kt:318]: Added a TODO

File: AnkiDroid/src/main/java/com/ichi2/compat/CompatV24.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/compat/CompatV24.kt]: This is great, thank you @Arthur-Milchior 

I've implemented it

**Comment by david-allison** [comment]
Force pushed a change to `AndroidManifest.xml` regarding drag & drop untested

**Review by mikehardy — COMMENTED** [review]
LGTM, cleaner commit setup, as hoped for with suggestion for BaseCompat (thanks Arthur) and splitting the shortcut stuff (thanks David)

**Comment by mikehardy** [comment]
It's still technically my PR so I can't +1 it but I'm +1

**Comment by david-allison** [comment]
Lint failure, mayve from my Manifest change. I don't see how

```
  Lint found 272 errors, 0 warnings. First failure:
  
  /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/res/values-ca/01-core.xml:59: Error: For locale "ca" (Catalan) the following quantity should also be defined: many [MissingQuantity]
      <plurals name="reviewer_window_title">
      ^
```

**Comment by mikehardy** [comment]
I think between the 3 of us we've all actually contributed to it and all ideas are ingested now, with CI green. I've re-read it, looks good to go

**Pull request merged**